src/main/java/com/google/sps/servlets/UploadServlet.java

```diff
@@ -0,0 +1,150 @@
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.nio.file.Files;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageOptions;
+import java.nio.file.Paths;
+
+/** Servlet for uploading files. */
+@WebServlet("/upload")
+@MultipartConfig
+public class UploadServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
+    //  Get all the Dataset Entities
+    Query query = new Query("MetaData").addSort("timestamp", SortDirection.ASCENDING);
+
+    /* Get current user logged in to webapp */
+    UserService userService = UserServiceFactory.getUserService();
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    /*  Get metadata (for comparison and eventual storage) */
+    String userEmail = userService.getCurrentUser().getEmail();
+    String datasetName = request.getParameter("dataset-namer");
+    long timestamp = System.currentTimeMillis();
+
+    // Loop to see if that user already exists
+    boolean datasetPresent = false;
+    boolean userPresent = false;
+    boolean newDataset = true;  // Assume this is a new dataset being created
+    for (Entity entity : results.asIterable()) {
+      String email = (String) entity.getProperty("user-email");
+      String dataset = (String) entity.getProperty("dataset-name");
+
+      if (email.equals(userEmail)) {
+        userPresent = true;
+        if (dataset.equals(datasetName)) {
+          // Do nothing, use this existing dataset
+          datasetPresent = true;
+          newDataset = false;
+        }
+      }
+    }
+    
+    // If this user already has that dataset present
+    if (userPresent && datasetPresent) {
+      response.sendRedirect("/invalid.html");
+    }
+
+
+    if (!datasetPresent) {
+      /* Upload images from form to GCP bucket */
+
+      int imageCount = 0;
+
+      // The ID of your GCP project
+      String projectId = "step-2020-johndallard";
+
+      // The ID of your GCS bucket
+      String bucketName = "embedding-visualizer-bucket";
+
+      Storage storage = StorageOptions.newBuilder().setProjectId(projectId).build().getService();
+
+      // Create new directory within user's subdirectory for this dataset
+      String userDirectory = userEmail + "/";
+      String newDatasetDir = userDirectory + datasetName + "/";
+      String userImagesDir = newDatasetDir + "original_images/";
+      String userThumbnailsDir = newDatasetDir + "thumbnails/";
+      String userSpritesheetDir = newDatasetDir + "spritesheets/";
+      String userEmbeddingsDir = newDatasetDir + "embeddings/";
+      String[] directories = {userImagesDir, userThumbnailsDir, userSpritesheetDir, userEmbeddingsDir};
```

Comment:
`ImmutableList<String>`
(Can you use Guava?)

Reply:
Ack. Changed!